python/ray/train/v2/trainer.py

```diff
@@ -0,0 +1,327 @@
+        Returns:
+            A Result object containing the training result.
+        """
+        trainable = self.as_trainable()
+
+        # Copied from initial prototyping.
+        # TODO: Replace with Tuner.
+        analysis = tune.run(run_or_experiment=trainable, **self.run_config)
+
+        assert len(analysis.trials) == 1
+
+        trial = analysis.trials[0]
+
+        result = Result(
+            metrics=trial.last_result,
+            checkpoint=Checkpoint.from_directory(trial.checkpoint.value)
+            if trial.checkpoint.value
+            else None,
+        )
+
+        return result
+
+    def _override_attributes_with_config(self, config: Dict):
+        """Overrides attributes of the Trainer with values in ``config``.
+
+        This is needed to allow attributes of Trainer to be tuned as
+        hyperparameters.
+
+        This method does a deep update of dictionaries and dataclasses
+
+        Args:
+            config (Dict): A dictionary to update attributes with.
+        """
+        for key in list(config.keys()):
+            if hasattr(self, key):
+                current_attribute = getattr(self, key)
+                if isinstance(current_attribute, dict):
+                    # Do a deep update of the dictionary.
+                    deep_update(current_attribute, config[key], new_keys_allowed=True)
+                elif dataclasses.is_dataclass(current_attribute):
+                    if type(config[key]) == type(current_attribute):
+                        # If the hyperparam is already passed in as a dataclass
+                        # Then just directly set it.
+                        setattr(self, key, config[key])
+                    elif isinstance(config[key], dict):
+                        # If the hyperparam is passed in as a dict, then update
+                        # each attribute of the dataclass directly.
+                        for inner_key, inner_value in config[key].items():
+                            if hasattr(current_attribute, inner_key):
+                                dataclass_field = getattr(current_attribute, inner_key)
+                                if isinstance(dataclass_field, dict):
+                                    # Do a deep update.
+                                    deep_update(
+                                        dataclass_field,
+                                        inner_value,
+                                        new_keys_allowed=True,
+                                    )
+                                else:
+                                    setattr(current_attribute, inner_key, inner_value)
+                else:
+                    # Don't do a deep update and directly set the attribute
+                    # to the value in config.
+                    setattr(self, key, config[key])
+                # Remove the key from the dict.
+                del config[key]
+
+
+@DeveloperAPI
+class DataParallelFunctionTrainer(Trainer):
```

Comment:
Can you move this into a separate file called `data_parallel_trainer.py`?

Reply:
Yep will do!